### ISSUE TITLE
AdvancedIncSubtensor1 increment broadcasting

### ISSUE THREAD

**Issue by sisp** [Theano/Theano#2811]
This is an additional case I mentioned [here](https://github.com/Theano/Theano/pull/2521#issuecomment-75554822) but didn't continue to focus on in that PR. The code in that PR basically showed a linear regression model with categorical inputs whose error is computed for a single example. The batch version would look like this
```
W = T.vector('W')
i = T.matrix('i', dtype='int64')
t = T.vector('t')
z = T.sqr(t - W[i].sum(axis=1)).mean()
g = theano.grad(z, W)
f = theano.function([W, i, t], g)
theano.printing.pydotprint(f, '/tmp/theano.png', format='png')
```
giving the following graph to compute the gradient of `W`:
![theano](https://cloud.githubusercontent.com/assets/2206639/7278583/b0e3d67c-e916-11e4-8925-342e136862b1.png)
As you can see there is still an `Alloc` op (Alloc id=19) followed by a `Reshape` op. This `Alloc` actually acts as a repeat op in order to replicate the elements of the incoming vector. But we don't actually need the replicated values. Just the fact that we flatten that resulting matrix so that is shape matches the vector that we add it to requires the explicit replication of the values. In NumPy I could imagine to use stride tricks (`strided_as` to be specific) to create a "virtual" dimension with the correct shape and strides=0 and use its flat view using the `flat` attribute to give me a 1D iterator. I'd save the cost of creating a copy with repeated elements.

Now, I'm not sure if it's possible to do something similar to avoid "Alloc id=19". A 1D iterator is read-only and has other properties than a regular ndarray, so it seems this approach is not compatible with the current Theano types. Does anyone see a way to get rid of that Alloc?

**Comment by lamblin** [comment]
Indeed, in some cases, `Reshape(Alloc(...))` can just be replaced with `Alloc` of the new shape (when the value allocated is scalar, for instance, or when the shapes have the right structure). It is not always possible.
Then the `Alloc` in input to `AdvancedIncSubtensor1` could be replaced by a `Dimshuffle` (or nothing) if it is supported.
